app/code/Magento/Catalog/Model/Product/Link/SaveHandler.php

```diff
@@ -55,17 +64,51 @@ public function execute($entityType, $entity)
     {
         $link = $entity->getData($this->metadataPool->getMetadata($entityType)->getLinkField());
         if ($this->linkResource->hasProductLinks($link)) {
-            /** @var \Magento\Catalog\Api\Data\ProductInterface $entity*/
+            /** @var \Magento\Catalog\Api\Data\ProductInterface $entity */
             foreach ($this->productLinkRepository->getList($entity) as $link) {
                 $this->productLinkRepository->delete($link);
             }
         }
         $productLinks = $entity->getProductLinks();
+
+        // Build links per type
+        $linksByType = [];
+        foreach ($productLinks as $link) {
```

Comment:
After your last refactoring, we don't need the variable `$productLinks` in line 72 anymore. Call `$entity->getProductLinks()` directly in line 72. By doing this, we adhere to the best practice that a variable shouldn't be changed anymore after initialisation.